launch_ros/launch_ros/actions/node.py

```diff
@@ -342,6 +383,10 @@ def _perform_substitutions(self, context: LaunchContext) -> None:
             raise
         self.__final_node_name = prefix_namespace(
             self.__expanded_node_namespace, self.__expanded_node_name)
+
+        for extension in self.__extensions.values():
+            self.cmd.extend(extension.command_extension(context))
```

Comment:
@gbiggs why a method to specifically extend the command? Isn't it possible to do so (even resolving substitutions) upon extension execution?

Reply:
I can't remember specifically, but I suspect I was following the existing pattern of the `Node` class, which splits up building the command and executing it. Possibly it does things that way for debug output?

Reply:
Hmm, neither `NodeExtension.command_extension` nor `NodeExtension.execute` deal with execution. Both configure the command line, in direct and indirect ways respectively.

Reply:
[`SecurityNodeActionExtension.execute`](https://github.com/osrf/ros2launch_security/pull/1/files#diff-d16580b357825167e86092c99e3945c9d131a8d1b8a4b4cf37de482298cef574R40-R44) could fetch an enclave and extend the command line w/o the extra step.

Reply:
It's not just fetching an enclave, it's doing stuff that has side effects. In the case of the security extension, it's actually creating and using files on disc. There's also a need to add an option to the `ros2 launch` command line for the user to use, which is why there is the separate `command_extension` method. The function documentation for that was wrong; I've fixed it in c2b2b7f.

Reply:
Hmm, I'm a bit confused. From what I can see in this patch, `NodeActionExtension.command_extension` is not used by `ros2 launch` directly, but within `Node._perform_substitutions` to extend `Node.cmd`. In which case, my point is that both `NodeActionExtension.command` and `NodeActionExtension.pre_execute` happen on `Action.execute`. There's no configuration stage in between. An indirection through context locals would make sense if there was a context global to override (or not).

Anyhow, I won't block on this. But we do need that documentation sorted out.

Reply:
OK, I was the one confused. Sorry about that. I was getting it mixed up with the `ros2 launch` command line flag adding, which is handled by the options extension.

I looked over the code and I agree that it is possible to combine the two functions. I can make an argument against combining them that the current way follows the structure of the `Node` action in that it partitions preparing the command line from preparing ROS-specific arguments and executing the command, but it's a weak argument. So I went ahead and changed it to match your suggestion in da09c67.